### PR TITLE
Add preferences for chePlugin components in Devfile

### PR DESCRIPTION
**Pull request by sleshchenko**
### What does this PR do?
This PR adds preferences for chePlugin components in Devfile.
As an example of Devfile where this functionality is used:
<details>
<summary>Devfile with configured typesctipt plugin</summary>

```yaml
specVersion: 0.0.1
name: cheTheia-in-che
projects:
  - name: devfiles
    source:
      type: git
      location: 'https://github.com/sunix/devfiles.git'
components:
  - 
    alias: theia-editor
    id: eclipse/che-theia/next
    type: cheEditor
  - 
    id: che-incubator/typescript/1.30.2
    type: chePlugin
    memoryLimit: 2048M
    preferences:
       java.jdt.ls.vmargs: '-noverify -Xmx1G -XX:+UseG1GC -XX:+UseStringDeduplication'
       java.home: '/home/user/jdk11'
commands:
  - name: 'say Hello'
    actions:
      - type: exec
        component: che-dev
        command: >
                echo "Hello!"
```

</details>

The same feature can be used with workspace config, `plugin.{PLUGIN_PUBLISHER}/{PLUGIN_NAME}.preference.{PREFERENCE_NAME}` workspace config attribute is used.
<details>
<summary>Workspace config with configured plugin</summary>

```json
{
    "projects": [
        {
            "source": {
                "location": "https://github.com/sunix/devfiles.git",
                "type": "git",
                "parameters": {}
            },
            "name": "devfiles",
            "path": "/devfiles"
        }
    ],
    "defaultEnv": "default",
    "name": "cheTheia-in-che",
    "attributes": {
        "editor": "eclipse/che-theia/next",
        "editorComponentAlias": "theia-editor",
        "plugin.che-incubator/typescript.preference.java.jdt.ls.vmargs": "-noverify -Xmx1G -XX:+UseG1GC -XX:+UseStringDeduplication",
        "plugins": "eclipse/che-machine-exec-plugin/0.0.1,che-incubator/typescript/1.30.2",
        "sidecar.che-incubator/typescript.memory_limit": "2048M",
        "plugin.che-incubator/typescript.preference.java.home": "/home/user/jdk11"
    }
}
```

</details>

There is a docker image `sleshchenko/che-server:plugin-prefs` that contains PR's changes. It may be used for testing.

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/12992

#### Release Notes
Added an ability to specify preferences for plugins in Devfile and Workspace Config.

#### Docs PR
Docs changes are included in the PR

### CODE REVIEW

**Review by skabashnyuk — APPROVED** [review]

**Comment by sleshchenko** [comment]
ci-test

**Comment by sleshchenko** [comment]
ci-test

**Comment by dmytro-ndp** [comment]
ci-test

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json [hunk 1]
```diff
@@ -191,6 +188,62 @@
...
+                    ]
+                  }
+                }
+              },
+              "then": {
+                "required": [
+                  "id"
+                ],
+                "additionalProperties": false,
+                "properties": {
```
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json:206]: Wouldn't it be useful to add the prefs also to the editor conceivably?

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json [hunk 2]
```diff
@@ -162,10 +162,7 @@
                 }
               },
               "then": {
-                "required": [
```
Comment (metlos) [inline wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json]: Could you explain the reason why this `required` declaration can no longer be here? Also, why does the `id` have to be specified for both the editor and plugin when we have this common declarations block here?

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Component.java [hunk 3]
```diff
@@ -28,6 +28,12 @@
   /** Returns the plugin/editor FQN. Is mandatory only for cheEditor/chePlugin components types. */
   String getId();
 
+  /**
+   * Returns the preferences of the plugin. Example value of preference: `java.home:
```
Comment (metlos) [inline core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Component.java]: Backticks are not supported in javadoc afaik. Use `{@code}` for the intended effect.

**Review by metlos — COMMENTED** [review]
Some minor comments but otherwise good!

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json:206]: I know nothing about preferences for Theia plugins but I asked @akurinnoy and he did not see sense in it.
@evidolob @l0rd @benoitf Do you think preferences can be useful for cheEditor component as well?

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json (hunk 2 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json]: Good catch. `required` can still be here. Changed

File: core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Component.java (hunk 3 above)
Comment (sleshchenko) [reply core/che-core-api-model/src/main/java/org/eclipse/che/api/core/model/workspace/devfile/Component.java]: Used {@code}

**Comment by SkorikSergey** [comment]
Selenium tests execution on Eclipse Che Multiuser on OCP (https://ci.codenvycorp.com/job/che-pullrequests-test-ocp/1801//Selenium_20tests_20report/) doesn't show any regression against this Pull Request.

**Review by l0rd — APPROVED** [review]

File: wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json (hunk 1 above)
Comment (sleshchenko) [reply wsmaster/che-core-api-devfile/src/main/resources/schema/devfile.json:206]: It may be added later if needed

**Comment by sleshchenko** [comment]
Rebased onto master branch without any changes in PR

**Comment by sleshchenko** [comment]
ci-build

**Pull request merged**